interpreter_book/playground3/monkey/evaluator/evaluator_add_panic_test.go

```diff
@@ -115,6 +116,7 @@ func checkPanic(input string, ast *ast.Program, hasParseErrors bool, t *testing.
 	env := object.NewEnvironment()
 	defer func() { // idea from https://golang.org/doc/effective_go#recover
 		if err := recover(); err != nil {
+			t.Error(string(debug.Stack()))
```

Comment:
I am not sure I remember if I was there when we talked about this. Is it only to help with better understanding what (and where) went wrong in the parser?

Reply:
No, you weren't there when it was added, it was a suggestion by @sleepypioneer  in the _one_ meeting you couldn't attend :-)

The deferred function is needed since many of the additional tests involve inputs for which the **evaluator** panics. Since we want to test several cases, we don't want our whole test to be killed by the first runtime exception, but want it to continue going through the rest of the test cases. 

In the code beneath the added line, the error message is printed, which gives us a reason **why** the evaluator panicked, 
whereas the added line additionally gives us information on the stack trace to help us debug **where** the evaluator panicked and what function calls led to calling this piece of code.